### PR TITLE
Kraken : Using Astar for direct path

### PR DESCRIPTION
**Pull request by AurelienLP**
This PR aims to :

- [ ] Creating a class PathFinder.
- [ ] Moving all the code related to the old PathFinder structure from street_network.h and street_network.cpp to path_finder.h and path_finder.cpp.
- [ ] Creating two classes DijkstraPathFinder and AstarPathFinder inherited from PathFinder which is abstract.
- [ ] Gathering the visitors for the research in visitor.h and using proper aliases.
- [ ] Fighting against huge templates.

### CODE REVIEW

File: source/georef/dijkstra_path_finder.cpp [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+
+    return nearest_edge.first;
+}
+
+/**
+  The _DEBUG_DIJKSTRA_QUANTUM_ activate at compil time some dump used in quantum to analyze
+  the street network.
+  WARNING, it will slow A LOT the djisktra and must be used only for debug
+  */
+#ifdef _DEBUG_DIJKSTRA_QUANTUM_
```
Comment (kinnou02) [inline source/georef/dijkstra_path_finder.cpp]: does it still build? if not we may want to delete it

File: source/georef/astar_path_finder.cpp [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+ * Warning, it modifies the distances and the predecessors
+ * Warning 2, the target node is NOT the destination !!!!
+ **/
+void AstarPathFinder::astar(const vertex_t source,
+                            const vertex_t target,
+                            const astar_distance_heuristic& heuristic,
+                            const astar_distance_or_target_visitor& visitor) {
+    // Note: the predecessors have been updated in init
+
+    // Fill color map in white before dijkstra
```
Comment (kinnou02) [inline source/georef/astar_path_finder.cpp]:
```suggestion
    // Fill color map in white before astar
```

**Review by kinnou02 — COMMENTED** [review]
It would be nice to have some test on the astar

File: source/georef/dijkstra_path_finder.cpp (hunk 1 above)
Comment (AurelienLP) [reply source/georef/dijkstra_path_finder.cpp]: No idea, but I am not against removing it

**Comment by kinnou02** [comment]
Jenkins, test this please.

**Review by kinnou02 — APPROVED** [review]

**Comment by AurelienLP** [comment]
Jenkins, test this please.

**Comment by AurelienLP** [comment]
Jenkins, test this please.

**Review by benoit-bst — APPROVED** [review]

File: source/georef/astar_path_finder.cpp [hunk 3]
```diff
@@ -72,7 +73,7 @@ void AstarPathFinder::astar(const vertex_t source,
     auto weight_map = boost::get(&Edge::duration, geo_ref.graph);
     auto combiner = SpeedDistanceCombiner(speed_factor);
 
-    astar_shortest_paths_no_init_with_heap(g, &source, &target, heuristic, visitor, weight_map, combiner);
+    astar_shortest_paths_no_init_with_heap(g, vertex.cbegin(), vertex.cend(), heuristic, visitor, weight_map, combiner);
```
Comment (kinnou02) [inline source/georef/astar_path_finder.cpp]: A comment explaining what happenned would help later

**Review by kinnou02 — APPROVED** [review]

File: source/georef/astar_path_finder.cpp [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+              astar_distance_heuristic(geo_ref.graph, destinations.back(), 1. / (default_speed[mode] * speed_factor)),
+              astar_distance_or_target_visitor(radius, distances, destinations));
+    } catch (DestinationFound&) {
+    }
+}
+
+/**
+ * Launch an astar without initializing the data structure
+ * Warning, it modifies the distances and the predecessors
+ * Warning 2, the target node is NOT the destination !!!!
```
Comment (nonifier) [inline source/georef/astar_path_finder.cpp]: If not the destination, what it is then ? :)

File: source/georef/astar_path_finder.cpp [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+ **/
+void AstarPathFinder::astar(const vertex_t source,
+                            const vertex_t target,
+                            const astar_distance_heuristic& heuristic,
+                            const astar_distance_or_target_visitor& visitor) {
+    // Note: the predecessors have been updated in init
+    std::array<georef::vertex_t, 2> vertex{{source, target}};
+
+    // Fill color map in white before dijkstra
```
Comment (nonifier) [inline source/georef/astar_path_finder.cpp]:
```suggestion
    // Fill color map in white before A*
```

File: source/georef/astar_path_finder.h [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+struct astar_distance_heuristic : public boost::astar_heuristic<Graph, navitia::seconds> {
+    const Graph& g;
+    const type::GeographicalCoord& dest_coord;
+    const double inv_speed;
+
+    astar_distance_heuristic(const Graph& graph, const vertex_t& destination, const double inv_speed)
+        : g(graph), dest_coord(graph[destination].coord), inv_speed(inv_speed) {}
+
+    navitia::seconds operator()(vertex_t v) {
```
Comment (nonifier) [inline source/georef/astar_path_finder.h]: you might want to pass a `const vertex_t&` here ?

File: source/georef/dijkstra_shortest_paths_with_heap.h [hunk 7]
```diff
@@ -30,7 +30,7 @@ www.navitia.io
 
 #include <boost/graph/dijkstra_shortest_paths.hpp>
 
-namespace boost {
+namespace navitia {
```
Comment (nonifier) [inline source/georef/dijkstra_shortest_paths_with_heap.h:33]: :heart_decoration:

File: source/georef/path_finder.cpp [hunk 8]
```diff
@@ -0,0 +1,517 @@
...
+    }
+    if (coords_to_add.empty()) {
+        coords_to_add.push_back(projection.projected);
+    }
+
+    auto duration = crow_fly_duration(projection.distances[d]);
+    // we need to update the total length
+    p.duration += duration;
+
+    // we aither add the starting coordinate to the first path item or create a new path item if it was another way
```
Comment (nonifier) [inline source/georef/path_finder.cpp]:
```suggestion
    // we either add the starting coordinate to the first path item or create a new path item if it was another way
```

File: source/georef/path_finder.cpp [hunk 9]
```diff
@@ -0,0 +1,517 @@
...
+}
+
+void PathFinder::add_custom_projections_to_path(Path& p,
+                                                bool append_to_begin,
+                                                const ProjectionData& projection,
+                                                ProjectionData::Direction d) const {
+    auto item_to_update = [append_to_begin](Path& p) -> PathItem& {
+        return (append_to_begin ? p.path_items.front() : p.path_items.back());
+    };
+    auto add_in_path = [append_to_begin](Path& p, const PathItem& item) {
```
Comment (nonifier) [inline source/georef/path_finder.cpp:213]: I don't know if it would make things simpler by using `std::front_inserter` or `std::back_inserter` with something like: 
```cpp
auto path_inserter = append_to_begin? std::front_inserter(p.path_tiems) : std::back_inserter(p.path_tiems);
[...]
path_inserter = item;
```

In a similar way, you could have to update your items:
```cpp
auto item_to_update = append_to_begin ? 
                     [&](){ return p.path_items.front(); } : 
                     [&](){ return p.path_items.back(); } ;
[...]
item_to_update().way_id 
```

Those are just ideas to make things nicer, feel free to disregard.
#ModernizeMyCpp

File: source/georef/path_finder.cpp [hunk 10]
```diff
@@ -0,0 +1,517 @@
...
+        if (!p.path_items.empty() && item_to_update(p).way_idx == type::invalid_idx) {
+            item_to_update(p).way_idx = start_edge.way_idx;
+            item_to_update(p).duration += duration;
+        } else {
+            PathItem item;
+            item.way_idx = start_edge.way_idx;
+            item.duration = duration;
+
+            if (!p.path_items.empty()) {
+                // still complexifying stuff... TODO: simplify this
```
Comment (nonifier) [inline source/georef/path_finder.cpp:256]: Maybe start with extracting the switch case into a separate function :)
And if you're not familiar with the SOLID principles, have a look at the first one - https://en.wikipedia.org/wiki/SOLID
TL;DR -
- a class has one responsibility
- a function does one thing

**Review by nonifier — APPROVED** [review]

File: source/georef/astar_path_finder.h (hunk 6 above)
Comment (xlqian) [reply source/georef/astar_path_finder.h]: and can it be a const operator?

File: source/georef/dijkstra_path_finder.cpp [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+                    distance_to_entry_point[sp_idx] = duration;
+                }
+            }
+        }
+    }
+    // case 2 : start coord is an edge (dijkstra)
+    else {
+        std::vector<routing::SpIdx> dest_sp_idx;
+        for (const auto& e : elements) {
+            dest_sp_idx.push_back(routing::SpIdx{e.first});
```
Comment (xlqian) [inline source/georef/dijkstra_path_finder.cpp]: you can use emplace_back

File: source/georef/path_finder.cpp (hunk 10 above)
Comment (AurelienLP) [reply source/georef/path_finder.cpp:256]: Is `get_transportation_mode_item_to_update` for the function containing the switch good enough for you ?

File: source/georef/dijkstra_path_finder.cpp [hunk 12]
```diff
@@ -169,7 +169,7 @@ routing::map_stop_point_duration DijkstraPathFinder::find_nearest_stop_points(
     else {
         std::vector<routing::SpIdx> dest_sp_idx;
         for (const auto& e : elements) {
-            dest_sp_idx.push_back(routing::SpIdx{e.first});
+            dest_sp_idx.emplace_back(routing::SpIdx{e.first});
```
Comment (benoit-bst) [inline source/georef/dijkstra_path_finder.cpp:172]: :+1:

**Review by nonifier — APPROVED** [review]

**Review by benoit-bst — APPROVED** [review]

File: source/georef/dijkstra_shortest_paths_with_heap.h (hunk 7 above)
Comment (pbougue) [reply source/georef/dijkstra_shortest_paths_with_heap.h:33]: I think we left that namespace because the goal was to make a PR to boost with that.

File: source/georef/path_finder.h [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+    virtual ~PathFinder() = 0;
+
+    /**
+     *  Update the structure for a given starting point and transportation mode
+     *  The init HAS to be called before any other methods
+     */
+    void init(const type::GeographicalCoord& start_coord, nt::Mode_e mode, const float speed_factor);
+
+    // return the path from the starting point to the target. the target has to have been previously visited.
+    Path get_path(type::idx_t idx);
```
Comment (pbougue) [inline source/georef/path_finder.h:157]: This method is only used in Astar, so it should probably move there.
Also not fan of the naming as it "shadows" the other `get_path`.

File: source/georef/path_finder.h [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+
+    // Add the starting point projection the the path. Add a new way if needed
+    void add_projections_to_path(Path& p, bool append_to_begin) const;
+
+    // shouldn't be used outside of class apart from tests
+    Path get_path(const ProjectionData& target,
+                  const std::pair<navitia::time_duration, ProjectionData::Direction>& nearest_edge);
+
+    // find the nearest vertex from the projection. return the distance to this vertex and the vertex
+    std::pair<navitia::time_duration, ProjectionData::Direction> find_nearest_vertex(const ProjectionData& target,
```
Comment (pbougue) [inline source/georef/path_finder.h:167]: This is currently the only method that needs to be public because it's used outside of tests or child classes.

Depending on the use of it, it feels like composition could be a clearer option than inheritance, as the children are not interchangeable (maybe they should be, and father class would define an interface then?)

File: source/georef/path_finder.h [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+    void init(const type::GeographicalCoord& start_coord, nt::Mode_e mode, const float speed_factor);
+
+    // return the path from the starting point to the target. the target has to have been previously visited.
+    Path get_path(type::idx_t idx);
+
+    // Add the starting point projection the the path. Add a new way if needed
+    void add_projections_to_path(Path& p, bool append_to_begin) const;
+
+    // shouldn't be used outside of class apart from tests
+    Path get_path(const ProjectionData& target,
```
Comment (pbougue) [inline source/georef/path_finder.h:163]: This is used in tests, but we should probably move it to protected section (and friend/inherit for tests).

File: source/georef/path_finder.h [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+    void add_custom_projections_to_path(Path& p,
+                                        bool append_to_begin,
+                                        const ProjectionData& projection,
+                                        ProjectionData::Direction d) const;
+
+    // Build a path with a destination and the predecessors list
+    Path build_path(vertex_t best_destination) const;
+};
+
+bool is_projected_on_same_edge(const ProjectionData& p1, const ProjectionData& p2);
```
Comment (pbougue) [inline source/georef/path_finder.h:193]: This could stay static to avoid unnecessary inclusion.

File: source/georef/astar_path_finder.cpp [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+void AstarPathFinder::astar(const std::array<georef::vertex_t, 2>& origin_vertexes,
+                            const astar_distance_heuristic& heuristic,
+                            const astar_distance_or_target_visitor& visitor) {
+    // Note: the predecessors have been updated in init
+
+    // Fill color map in white before A*
+    std::fill(color.data.get(), color.data.get() + (color.n + color.elements_per_char - 1) / color.elements_per_char,
+              0);
+
+    // we filter the graph to only use certain mean of transport
```
Comment (pbougue) [inline source/georef/astar_path_finder.cpp:67]: Not sure: Can't you do that at init? To avoid multiple filtering calls?

**Review by pbougue — COMMENTED** [review]
Mostly for memory: some of these remarks are not introduced by this PR, and I did not dig much into it with all that code moving :dizzy_face: :smile: 
But in case :truck:-team wants to cleanup, here are some ideas.

The point on filtering might improve perf, though.
